### PR TITLE
Implement s3 filesigner

### PR DESCRIPTION
**Pull request by matar993**

### CODE REVIEW

File: server/src/main/java/io/whitefox/core/services/DeltaSharesServiceImpl.java [hunk 1]
```diff
@@ -121,13 +121,19 @@ public Optional<ContentAndToken<List<SharedTable>>> listTablesOfShare(
...
+        .orElseThrow(() -> new TableNotFound(String.format(
+            "Table %s not found with share %s and schema %s", tableName, share, schema)));
+
+    return Optional.ofNullable(tableLoader.loadTable(sharedTable))
         .map(dst -> dst.queryTable(queryRequest))
         .map(result -> new ReadTableResult(
             result.protocol(),
             result.metadata(),
-            result.other().stream().map(signer::sign).collect(Collectors.toList())));
+            result.other().stream()
```
Comment (tmnd1991) [inline server/src/main/java/io/whitefox/core/services/DeltaSharesServiceImpl.java:134]: why do you need a new fileSigner for each element?

File: server/src/main/java/io/whitefox/core/services/S3FileSigner.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+  private String stripPrefix(String string, String prefix) {
+    if (string.startsWith(prefix)) {
+      return string.substring(prefix.length());
+    } else return string;
+  }
+
+  private GetObjectPresignRequest getObjectPresignRequest(String bucketName, String keyName) {
+    return GetObjectPresignRequest.builder()
+        .signatureDuration(Duration.ofMinutes(60)) // TODO
```
Comment (tmnd1991) [inline server/src/main/java/io/whitefox/core/services/S3FileSigner.java:50]: Let's make it configurable from the storage object (default 60 minutes is fine)

File: server/src/main/resources/application.properties [hunk 3]
```diff
@@ -1,2 +1,5 @@
 quarkus.http.test-port=8083
-io.delta.sharing.api.server.defaultMaxResults=10
\ No newline at end of file
+io.delta.sharing.api.server.defaultMaxResults=10
+
+# aws
+provider.aws.region=eu-west-1
```
Comment (tmnd1991) [inline server/src/main/resources/application.properties:5]: this should be picked up from the storage object, not from application configuration.

File: server/src/native-test/java/io/whitefox/api/deltasharing/server/DeltaSharesApiImplIT.java [hunk 4]
```diff
@@ -3,15 +3,15 @@
...
+import io.quarkus.test.junit.QuarkusTest;
 import io.whitefox.OpenApiValidationFilter;
 import io.whitefox.api.deltasharing.encoders.DeltaPageTokenEncoder;
 import io.whitefox.core.services.ContentAndToken;
 import jakarta.ws.rs.core.Response;
 import java.nio.file.Paths;
 import org.junit.jupiter.api.Test;
 
-@QuarkusIntegrationTest
+@QuarkusTest
```
Comment (tmnd1991) [inline server/src/native-test/java/io/whitefox/api/deltasharing/server/DeltaSharesApiImplIT.java:14]: Why this change?

File: server/src/test/java/io/whitefox/api/deltasharing/S3Config.java [hunk 5]
```diff
@@ -0,0 +1,35 @@
+package io.whitefox.api.deltasharing;
+
+import jakarta.inject.Inject;
+import jakarta.inject.Singleton;
+import org.eclipse.microprofile.config.inject.ConfigProperty;
+
+@Singleton
+public class S3Config {
```
Comment (tmnd1991) [inline server/src/test/java/io/whitefox/api/deltasharing/S3Config.java:8]: Why do we need this?

File: server/src/test/java/io/whitefox/api/deltasharing/model/FileObjectFileWithoutPresignedUrl.java [hunk 6]
```diff
@@ -0,0 +1,129 @@
+package io.whitefox.api.deltasharing.model;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import jakarta.validation.constraints.NotNull;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+
+public class FileObjectFileWithoutPresignedUrl {
```
Comment (tmnd1991) [inline server/src/test/java/io/whitefox/api/deltasharing/model/FileObjectFileWithoutPresignedUrl.java:9]: Can we make openapi-generator generate this and `server/src/test/java/io/whitefox/api/deltasharing/model/FileObjectWithoutPresignedUrl.java`?
actually they should already be there

File: server/src/test/java/io/whitefox/core/FakeFileSignerFactory.java [hunk 7]
```diff
@@ -0,0 +1,13 @@
+package io.whitefox.core;
+
+import io.whitefox.core.services.FileSigner;
+import io.whitefox.core.services.FileSignerFactory;
+import io.whitefox.core.services.NoOpSigner;
+
+public class FakeFileSignerFactory implements FileSignerFactory {
```
Comment (tmnd1991) [inline server/src/test/java/io/whitefox/core/FakeFileSignerFactory.java:7]: Probably you can build a single factory that returns both noop and s3 file signer based on the storage object passed it.

File: server/src/test/resources/application.properties [hunk 8]
```diff
@@ -1,2 +1,6 @@
 quarkus.http.test-port=8083
-io.delta.sharing.api.server.defaultMaxResults=10
\ No newline at end of file
+io.delta.sharing.api.server.defaultMaxResults=10
+
+provider.aws.region=eu-west-1
+provider.aws.accessKey=${APP_AWS_ACCESS_KEY_ID}
+provider.aws.secretKey=${APP_AWS_SECRET_ACCESS_KEY}
```
Comment (tmnd1991) [inline server/src/test/resources/application.properties:6]: why are these needed?

**Review by tmnd1991 — CHANGES_REQUESTED** [review]

File: server/src/native-test/java/io/whitefox/api/deltasharing/server/DeltaSharesApiImplIT.java (hunk 4 above)
Comment (matar993) [reply server/src/native-test/java/io/whitefox/api/deltasharing/server/DeltaSharesApiImplIT.java:14]: this is because this native test are fails during build.

File: server/src/test/java/io/whitefox/api/deltasharing/S3Config.java (hunk 5 above)
Comment (matar993) [reply server/src/test/java/io/whitefox/api/deltasharing/S3Config.java:8]: It's only for integration tests

File: server/src/test/resources/application.properties (hunk 8 above)
Comment (matar993) [reply server/src/test/resources/application.properties:6]: they are for tests only. I can highlight this

File: server/src/test/java/io/whitefox/api/deltasharing/model/FileObjectFileWithoutPresignedUrl.java (hunk 6 above)
Comment (matar993) [reply server/src/test/java/io/whitefox/api/deltasharing/model/FileObjectFileWithoutPresignedUrl.java:9]: I don't see this class in the protocol. I can see only the `FileObject` object (which includes the presigned url and the hashed id). The `FileObjectWithoutPresignedUrl` object has been defined to be used only in the testing phase, as it is not possible to control the presigned url and the hashed id. I could also decide to use the real `FileObject` object and ignore the presigned url and id (hashed). However, it seemed cleaner to me to define a class without those two fields, and make the assertion on the whole object.

File: server/build.gradle.kts [hunk 9]
```diff
@@ -34,6 +34,11 @@ dependencies {
     testImplementation("io.rest-assured:json-path")
     testImplementation("org.openapi4j:openapi-operation-validator:1.0.7")
     testImplementation("org.openapi4j:openapi-operation-restassured:1.0.7")
+
+    //AWS
+    implementation(platform("software.amazon.awssdk:bom:2.21.1"))
+    implementation("software.amazon.awssdk:s3")
```
[the remark this answers is not in the document]
Comment (matar993) [reply server/build.gradle.kts:40]: Unfortunately not

[the remark this answers is not in the document]
Comment (tmnd1991) [reply server/build.gradle.kts:40]: I guess Hadoop-aws imports aws sdk 1, would It be possible to import only aws sdk v1?

File: server/build.gradle.kts [hunk 10]
```diff
@@ -26,14 +26,20 @@ dependencies {
...
     testImplementation("io.quarkus:quarkus-junit5")
     testImplementation("io.rest-assured:rest-assured")
     testImplementation("io.rest-assured:json-path")
     testImplementation("org.openapi4j:openapi-operation-validator:1.0.7")
     testImplementation("org.openapi4j:openapi-operation-restassured:1.0.7")
+
+    //AWS
+    implementation(platform("software.amazon.awssdk:bom:2.21.1"))
+    implementation("software.amazon.awssdk:s3")
+    testImplementation(String.format("org.apache.hadoop:hadoop-aws:%s", hadoopVersion))
```
Comment (tmnd1991) [inline server/build.gradle.kts:41]: This should be a runtime dependency, not test one

File: server/src/main/java/io/whitefox/core/services/HadoopConfigBuilder.java [hunk 11]
```diff
@@ -0,0 +1,38 @@
+package io.whitefox.core.services;
+
+import io.whitefox.core.AwsCredentials;
+import io.whitefox.core.Storage;
+import io.whitefox.core.StorageProperties;
+import jakarta.enterprise.context.ApplicationScoped;
+import org.apache.hadoop.conf.Configuration;
+
+@ApplicationScoped
```
Comment (tmnd1991) [inline server/src/main/java/io/whitefox/core/services/HadoopConfigBuilder.java:9]: Are you sure we need this? No one injects this

File: server/src/test/java/io/whitefox/core/services/S3FileSignerTest.java [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+import io.quarkus.test.junit.QuarkusTest;
+import io.whitefox.api.deltasharing.S3TestConfig;
+import io.whitefox.core.*;
+import jakarta.inject.Inject;
+import java.util.Map;
+import java.util.Optional;
+import org.junit.jupiter.api.Test;
+import org.junit.platform.commons.util.StringUtils;
+
+@QuarkusTest
```
Comment (tmnd1991) [inline server/src/test/java/io/whitefox/core/services/S3FileSignerTest.java:14]: Do we need this to be a quarkus test?

**Review by tmnd1991 — CHANGES_REQUESTED** [review]

File: server/src/test/java/io/whitefox/api/deltasharing/model/FileObjectFileWithoutPresignedUrl.java (hunk 6 above)
Comment (tmnd1991) [reply server/src/test/java/io/whitefox/api/deltasharing/model/FileObjectFileWithoutPresignedUrl.java:9]: If it's needed only in tests, it's better like this ;) sorry I did not get it earlier

File: server/src/native-test/java/io/whitefox/api/deltasharing/server/DeltaSharesApiImplIT.java (hunk 4 above)
Comment (tmnd1991) [reply server/src/native-test/java/io/whitefox/api/deltasharing/server/DeltaSharesApiImplIT.java:14]: Well, this is not the way to solve it 😛

File: server/src/main/resources/application.properties (hunk 3 above)
Comment (tmnd1991) [inline server/src/main/resources/application.properties]: I think this too should be picked by the Storage object, not from a application wide conf

File: server/src/main/java/io/whitefox/core/services/S3FileSigner.java (hunk 2 above)
Comment (tmnd1991) [reply server/src/main/java/io/whitefox/core/services/S3FileSigner.java:50]: followup issue created

File: server/src/test/java/io/whitefox/core/services/S3FileSignerTest.java (hunk 12 above)
Comment (tmnd1991) [reply server/src/test/java/io/whitefox/core/services/S3FileSignerTest.java:14]: so we need it because of S3TestConfig, but is it worth to make the test 1-2 seconds slower only to get 3 environment variables?

Comment (matar993) [reply server/src/test/java/io/whitefox/core/services/S3FileSignerTest.java:14]: Yes because it is an integration test in that it signs a file on s3. it is an integration test because I need to initialize an `S3FileSigner`, to do that I need whitefox access keys, and to get them I need the application to be up.

On the other hand, if I turn it into a unit test by using fake credentials and mocking the `S3FileSigner`, I would not actually test anything.

File: server/src/main/java/io/whitefox/core/services/HadoopConfigBuilder.java (hunk 11 above)
Comment (matar993) [reply server/src/main/java/io/whitefox/core/services/HadoopConfigBuilder.java:9]: You're right

File: server/src/main/resources/application.properties (hunk 3 above)
Comment (matar993) [reply server/src/main/resources/application.properties]: I think you are seeing an outdated file.

Comment (matar993) [reply server/src/main/resources/application.properties]: Anyway I had just forgotten to remove it, in fact it is not used by anyone. The region, accessKey and secretKey are always retrieved from a storage object

File: server/src/main/java/io/whitefox/core/services/S3FileSigner.java [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+  @Override
+  public TableFile sign(TableFileToBeSigned s) {
+    URI absPath = URI.create(s.url());
+    String bucketName = absPath.getHost();
+    String keyName = stripPrefix(absPath.getPath(), "/");
+    PresignedGetObjectRequest presignedGetObjectRequest =
+        presigner.presignGetObject(getObjectPresignRequest(bucketName, keyName));
+    return new TableFile(
+        presignedGetObjectRequest.url().toString(),
+        tableFileIdHashFunction.hash(presignedGetObjectRequest.url().toString()),
```
Comment (tmnd1991) [inline server/src/main/java/io/whitefox/core/services/S3FileSigner.java:33]: this should be the hash of the original URL not the pre-signed one (that will change every time)

File: server/src/test/java/io/whitefox/core/services/S3FileSignerTest.java (hunk 12 above)
Comment (tmnd1991) [reply server/src/test/java/io/whitefox/core/services/S3FileSignerTest.java:14]: yes it's worth since this test bottleneck is I/O setting up quarkus is fast enough to make it not noticeable

**Comment by tmnd1991** [comment]
Now some tests fail locally if you don't provide environment variables:

- WHITEFOX_TEST_AWS_REGION
- WHITEFOX_TEST_AWS_SECRET_ACCESS_KEY
- WHITEFOX_TEST_AWS_ACCESS_KEY_ID

Please add some doc and links for developers so that they know how to obtain those, or even better, mark those tests so they can be later ignored if needed.

File: server/build.gradle.kts [hunk 14]
```diff
@@ -26,14 +26,20 @@ dependencies {
...
     // TEST
     testImplementation("io.quarkus:quarkus-junit5")
     testImplementation("io.rest-assured:rest-assured")
     testImplementation("io.rest-assured:json-path")
     testImplementation("org.openapi4j:openapi-operation-validator:1.0.7")
     testImplementation("org.openapi4j:openapi-operation-restassured:1.0.7")
+
+    //AWS
+    implementation("com.amazonaws:aws-java-sdk-bom:1.12.429")
+    implementation("com.amazonaws:aws-java-sdk-s3:1.12.470")
```
Comment (tmnd1991) [inline server/build.gradle.kts:40]:
```suggestion
    compileOnly("com.amazonaws:aws-java-sdk-bom:1.12.429")
    compileOnly("com.amazonaws:aws-java-sdk-s3:1.12.470")
```

**Review by tmnd1991 — APPROVED** [review]
